File: go/src/other/array/single-number-3.go
package array

func singleNumber3(nums []int) (res []int) {
	diff := 0
	for n := range nums {
		diff ^= n
	}

	diff &= -diff

	var result = make([]int, 2)
	result[0], result[1] = 0, 0

	for n := range nums {
		if (n & diff) == 0 {
			result[0] ^= n
		} else {
			result[1] ^= n
		}
	}
	return result
}

File: python/practices/array/single-number-3.py
def single_numbers(nums):
    diff = 0
    for n in nums:
        diff ^= n

    diff &= -diff

    rets = [0, 0]
    for n in nums:
        if (n & diff) == 0:
            rets[0] ^= n
        else:
            rets[1] ^= n

    return rets
